### PR TITLE
Expand capabilities of dealing with complex parameters

### PR DESCRIPTION
**Pull request by apuignav**
- [x] Implement `PolarComplexParameter`, working with tensors.
- [x] Add `conj` method to `*ComplexParameter`s.
- [x] Tests

### CODE REVIEW

**Comment by apuignav** [comment]
So far, I think this only works with floats as inputs, no tensors, but I need @mayou36 help for that.

**Comment by mayou36** [comment]
We could even try to make it better: the only difference between the angular and the complex is the constructor, right? So we can move everything from ComplexParamtere to BaseComplexParameter and inherit from that (so that `ComplexParameter` is basically an "empty" inherit, `ComplexPolarParameter` adds somethings in the init.

**Comment by apuignav** [comment]
Done! What do we gain with this, though?

**Comment by mayou36** [comment]
Hm, no, I have to think it through. Feels like somethings still odd

**Comment by apuignav** [comment]
OK, think about it, I'll write a couple of tests in the meantime.

File: zfit/core/parameter.py [hunk 1]
```diff
@@ -465,23 +465,51 @@ class ComposedParameter(BaseComposedParameter):
...
+
+    @staticmethod
+    def from_polar(name, abs, arg, dtype=ztypes.complex, **kwargs):
+        return ComplexParameter(name, tf.cast(tf.complex(abs*tf.math.cos(arg),
+                                                         abs*tf.math.sin(arg)),
+                                              dtype=dtype))
+
+    def conj(self):
+        return ComposedParameter('{}_conj'.format(self.name),
+                                 tf.math.conj(self))
```
Comment (mayou36) [inline zfit/core/parameter.py]: This should be a ComplexParameter as well, right?

And all the arguments should be propagated, e.g. floating=self.floating, dtype=...

File: zfit/core/parameter.py [hunk 2]
```diff
@@ -465,23 +465,51 @@ class ComposedParameter(BaseComposedParameter):
...
-        params = get_dependents(tensor=tensor, candidates=independend_params)
+        independent_params = tf.get_collection("zfit_independent")
+        params = get_dependents(tensor=tensor, candidates=independent_params)
         # params_init_op = [param.initializer for param in params]
         params = {p.name: p for p in params}
         # with tf.control_dependencies(params_init_op):
         super().__init__(params=params, initial_value=tensor, name=name, **kwargs)
 
 
 class ComplexParameter(BaseComposedParameter):
```
Comment (mayou36) [inline zfit/core/parameter.py]: Actually, why not just inherit from ComposedParameter? It *is* a composed parameter and reduces code dublication

File: zfit/core/parameter.py [hunk 3]
```diff
@@ -465,23 +465,51 @@ class ComposedParameter(BaseComposedParameter):
...
+        params = get_dependents(tensor=tensor, candidates=independent_params)
+        params = {p.name: p for p in params}
+        super().__init__(name=name, params=params, initial_value=tensor, **kwargs)
+
+    @staticmethod
+    def from_cartesian(name, real, imag, dtype=ztypes.complex, **kwargs):
+        return ComplexParameter(name, tf.cast(tf.complex(real, imag), dtype=dtype))
+
+    @staticmethod
+    def from_polar(name, abs, arg, dtype=ztypes.complex, **kwargs):
```
Comment (mayou36) [inline zfit/core/parameter.py]: abs -> mod?

so we have name consistency with the attribute and no clash with builtins

File: zfit/core/parameter.py [hunk 4]
```diff
@@ -465,23 +465,51 @@ class ComposedParameter(BaseComposedParameter):
...
+                                              dtype=dtype))
+
+    def conj(self):
+        return ComposedParameter('{}_conj'.format(self.name),
+                                 tf.math.conj(self))
+
+    @property
+    def real(self):
+        return tf.real(self.value())
+
```
Comment (mayou36) [inline zfit/core/parameter.py]: Just `self` is enough, remove the `.value()`. `value()` does a certain read, there would also be `read_value`. It's kinda unclear tough, but to be safe (and something for the new tf Variables to see how they do it/recommend ;) )

File: zfit/core/parameter.py (hunk 1 above)
Comment (apuignav) [reply zfit/core/parameter.py]: Ah yes, `floating` and `dtype`?

File: zfit/core/parameter.py (hunk 2 above)
Comment (apuignav) [reply zfit/core/parameter.py]: Oh, so true! Fixed!

File: zfit/core/parameter.py (hunk 1 above)
Comment (mayou36) [reply zfit/core/parameter.py]: Yes, that's it

File: zfit/core/parameter.py (hunk 3 above)
Comment (apuignav) [reply zfit/core/parameter.py]: OK!

File: zfit/core/parameter.py (hunk 4 above)
Comment (apuignav) [reply zfit/core/parameter.py]: OK!

File: zfit/core/parameter.py (hunk 2 above)
Comment (apuignav) [reply zfit/core/parameter.py]: Since `convert_to_tensor` has a hardcoded `dtype.float`, I updated `ComposedParameter`.

**Comment by mayou36** [comment]
Nice catch with the `ztf.convert_to_tensor` dtype!

I'll gonna look it through tomorrow, just to think through again but looks good (I can also resolve the conflicts)

**Comment by apuignav** [comment]
OK, I let you to fix conflicts and merge ;-)

**Comment by apuignav** [comment]
Thanks! That was really useful!

**Pull request merged**